src/core/thread/mle_router.cpp

```diff
@@ -3494,5 +3546,128 @@ ThreadError MleRouter::AppendPendingDataset(Message &aMessage)
     return error;
 }
 
+bool MleRouter::HasMinDowngradeNeighborRouters(void)
+{
+    return GetMinDowngradeNeighborRouters() >= kMinDowngradeNeighbors;
+}
+
+bool MleRouter::HasOneNeighborwithComparableConnectivity(const RouteTlv &aRoute, uint8_t aRouterId)
+{
+    uint8_t localLqi = 0;
+    uint8_t peerLqi = 0;
+    uint8_t routerCount = 0;
+
+    // process local neighbor routers
+    for (uint8_t i = 0; i < kMaxRouterId; i++)
+    {
+        if (i == mRouterId)
+        {
+            routerCount++;
+            continue;
+        }
+
+        // check if neighbor is valid
+        if (mRouters[i].mState == Neighbor::kStateValid)
+        {
+            // if peer is just neighbor
+            if (i == aRouterId)
+            {
+                routerCount++;
+                continue;
+            }
+
+            localLqi = mRouters[i].mLinkQualityOut;
+
+            if (localLqi > mRouters[i].mLinkInfo.GetLinkQuality(mMac.GetNoiseFloor()))
+            {
+                localLqi = mRouters[i].mLinkInfo.GetLinkQuality(mMac.GetNoiseFloor());
```

Comment:
nit: it would be more efficient initialize `localLqi` with Link Quality In and check against Link Quality Out.